List games

alexbanilean opened the pull request.

nicugnm commented on src/main/java/ro/unibuc/triplea/application/auth/dto/response/GameResponse.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+@Builder
+public class GameResponse {
+    private Integer gameSteamId;
```
gameSteamId can be nullable?

alexbanilean replied:
replaced with int, cannot be nullable

nicugnm commented on src/main/java/ro/unibuc/triplea/application/auth/web/GameController.java:
```diff
@@ -0,0 +1,49 @@
+    
+    private final GameService gameService;
+
+    @GetMapping("/get-all")
```
REST API should not contain verbs.

https://medium.com/@schneidenbach/restful-api-best-practices-and-common-pitfalls-7a83ba3763b5

alexbanilean replied:
renamed with ```games-list```

nicugnm commented on src/main/java/ro/unibuc/triplea/application/auth/web/GameController.java:
```diff
@@ -0,0 +1,49 @@
+        
+        Optional<GameResponse> game;
+
+        try {
```
Business logic should be into service. Create it's own method in a service class.

alexbanilean replied:
refactored code

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/auth/model/entity/Game.java:
```diff
@@ -0,0 +1,57 @@
+    @Column(name = "game_name", nullable = true)
+    private String gameName;
+
+    public Integer getId() {
```
Instead of generating getter and setters use lombok @Getter and @Setter. Also, you use @Data which contain getter and setters and toString (and others) so all of those can be deleted.

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/auth/model/entity/Game.java:
```diff
@@ -0,0 +1,57 @@
+    private Integer id;
+
+    @Column(name = "game_steam_id", nullable = false)
+    private Integer gameSteamId;
```
Why do we need id and gameSteamId. Can't be use gameSteamId as primary key?

vgrigoriu replied:
What if you later decide to extend this to games from other providers?

RaduADumitru replied:
Good question, then I think we can leave this code like this: leave id as the internal identifier for all games in our platform, and add columns for the id from each game provider, as needed

nicugnm replied:
Maybe a good approach (if we would have another providers) is to create a base class named Game and create another class SteamGame which will contain gameSteamId

alexbanilean replied:
Created SteamGame class that extends Game class

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/auth/service/GameService.java:
```diff
@@ -0,0 +1,31 @@
+    
+    private final GameRepository gameRepository;
+
+    @Autowired
```
Autowired can be avoided, also constructor. Use @RequiredArgsConstructor as used in many other files as injection

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/SecurityConfiguration.java:
```diff
@@ -26,7 +26,7 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .csrf(csrf -> csrf.disable())
                 .authorizeHttpRequests(auth -> auth
-                        .requestMatchers("/api/v1/auth/**", "/v3/api-docs/**", "/swagger-ui/**").permitAll()
+                        .requestMatchers("/api/v1/auth/**", "/v3/api-docs/**", "/swagger-ui/**", "/api/v1/games/**").permitAll()
```
/api/v1/games/** have access everyone. do we want that? i think an authenticated user (eventually based on role) should have access to those endpoints.

RaduADumitru replied:
Here I think we can allow everyone, since those endpoints are only for listing games. If we had a FE I'd imagine we could have this info on a main page available for everyone, including unauthenticated users. The next endpoints we make though, we should leave behind security (reviews/favourites)

alexbanilean replied:
It remains public as the team agreed.

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+    private final RestTemplate restTemplate;
+    private final SpringDataGameRepository springDataGameRepository;
+
+    public GameRepositoryImpl(RestTemplate restTemplate, SpringDataGameRepository springDataGameRepository) {
```
Instead of constructor, use @RequiredArgsConstructor.

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+        ResponseEntity<JsonNode> responseEntity = restTemplate.getForEntity(apiUrl, JsonNode.class);
+
+        if (responseEntity.getStatusCode().is2xxSuccessful()) {
+
```
remove this line

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+
+            return games;
+        } else {
+            System.err.println("Failed to fetch games from the external API. Status code: " + responseEntity.getStatusCode());
```
Instead of using system.out.println let's use @Slf4j or any other logger. System function should not be used as logging.

alexbanilean replied:
used slf4j

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public List<GameResponse> findAll(Optional<Integer> count) {
```
findAll is an improper name for this function because you provide a count as argument. Maybe another name should be fine. findAll is a proper name for a function that finds all names without needing a counting argument

alexbanilean replied:
renamed with findGames

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+                            break;
+                        }
+
+                        if (app.get("name").asText() != "") {
```
string should not be compared with !=. Use .equals() instead or .equalsIgnoreCase()

RaduADumitru replied:
To avoid NPEs we could also switch the order of the comparison, so that it won't try to find methods of a null object and cause an error:
`if !("".equals(app.get("name").asText()))`

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+                }
+            }
+        } else {
+            System.err.println("Failed to fetch games from the external API. Status code: " + responseEntity.getStatusCode());
```
Do not use System.err as logging. Use @Slf4j or another logging alternative (log4j , etc.)

alexbanilean replied:
used slf4j

nicugnm commented on src/main/java/ro/unibuc/triplea/infrastructure/game/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,132 @@
+                }
+            }
+        } else {
+            System.err.println("Failed to fetch games from the external API. Status code: " + responseEntity.getStatusCode());
```
Do not use System.err as logging. Use @slf4j or another logging alternative (log4j , etc.)

alexbanilean replied:
used slf4j

nicugnm commented:
For the moment, your implementation calls an external api from a service. That violates single responsability principle. A better approach would be to create a gateway package, an interface that's part of the domain and then extend that class in the infrastructure layer and the implementation should be responsible only with external api calls.

If team agree, actual implementation could work but could be improved in that way.

RaduADumitru commented:
> For the moment, your implementation calls an external api from a service. That violates single responsability principle. A better approach would be to create a gateway package, an interface that's part of the domain and then extend that class in the infrastructure layer and the implementation should be responsible only with external api calls.
> 
> If team agree, actual implementation could work but could be improved in this way.

I think that's a good approach since we could then also reuse the same code in other services. (We'd need to make similar calls for example for the favourite/reviews to check if a game with a steam id exists). However since this project doesn't have a large scope, it wouldn't be a bad idea to keep it the way it is now either. @alexbanilean I'd leave the decision up to you, you could add this if you have capacity, but if not, then I'd have no issue with leaving it like this

nicugnm commented:
Also, while there are errors, please proper exceptions and controller advice for the following implementations.

alexbanilean commented:
> For the moment, your implementation calls an external api from a service. That violates single responsability principle. A better approach would be to create a gateway package, an interface that's part of the domain and then extend that class in the infrastructure layer and the implementation should be responsible only with external api calls.
> 
> If team agree, actual implementation could work but could be improved in this way.

I created a gateway package and moved logic there in order to comply with SRP. There I added basic filtering for the external API data too.

alexbanilean commented:
> Also, while there are errors, please create proper exceptions and controller advice for the following implementations. You can take a look on auth examples.

I created custom exceptions as suggested.

nicugnm commented on src/main/java/ro/unibuc/triplea/application/games/steam/web/SteamGameController.java:
```diff
@@ -0,0 +1,44 @@
+        if (game.isPresent()) {
+            return ResponseEntity.ok(game.get());
+        } else {
+            throw new SteamGameNotFoundException("Steam game with identifier " + identifier + " not found");
```
Exception should not be treated in controller, move them to service.

alexbanilean replied:
fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/auth/exception/advice/AppWideExceptionHandler.java:
```diff
@@ -1,20 +1,29 @@
+    }
+
+    @ExceptionHandler(SteamGameValidateException.class)
+    public ResponseEntity<StandardResponse> handleValidationException(SteamGameValidateException e) {
```
This violates SRP. Move Games to it's own controller advice.

alexbanilean replied:
Fixed. Maybe a more specific name for this handler can be given, not ```AppWideExceptionHandler```.

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/auth/exception/advice/AppWideExceptionHandler.java:
```diff
@@ -1,20 +1,29 @@
-    @ExceptionHandler(Exception.class)
-    public ResponseEntity handleException(Exception e) {
-        return new ResponseEntity(new StandardResponse("500", "Error", e.getMessage()), HttpStatus.INTERNAL_SERVER_ERROR);
+    @ExceptionHandler(SteamGameNotFoundException.class)
```
This violates SRP. Move Games to it's own controller advice. (games)

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/games/steam/service/SteamGameService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private boolean isNumeric(String str) {
+        return str.matches("\\d+");
```
Move to an utility class.

alexbanilean replied:
Fixed

nicugnm commented on src/main/java/ro/unibuc/triplea/domain/games/steam/service/SteamGameService.java:
```diff
@@ -0,0 +1,53 @@
+        return str.matches("\\d+");
+    }
+    
+    private boolean isValidString(String str) {
```
Move to an utility class.

alexbanilean replied:
Fixed. Created an ```IdentifierUtil``` utility class.

nicugnm commented on src/main/java/ro/unibuc/triplea/application/games/steam/dto/response/SteamGameStandardResponse.java:
```diff
@@ -0,0 +1,14 @@
+public class SteamGameStandardResponse {
+    private String code;
+    private String message;
+    private Object data;
```
Why data is type of Object? object is an internal class and should not be used

alexbanilean replied:
Using ```Object``` allows flexibility in the type of data that can be stored in the ```data``` field, even though it sacrifices type safety. I took as model the ```StandardResponse``` class in auth implementation, and made a specific one for the Steam Game package. I modified the type now to String, but I think hanges should be made in auth package too.
